src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+        # with name 'ref(name.chunk-index)'
+        if metadata is None:
+            return None
+        properties = dict()
+        for k, v in metadata.items():
+            if isinstance(v, bytes):
+                b64 = base64.b64encode(v).decode()
+                value = Backend.VALUE_BASE64 + b64
+                if len(k+value) <= Backend.MAX_PROPERTY_SIZE:
+                    properties[k] = value
+                else:
+                    # create chunked metadata property
+                    chunks = [b64[i:i+Backend.MAX_REF_CHUNK_SIZE]
+                        for i in range(0, len(b64), Backend.MAX_REF_CHUNK_SIZE)]
+                    properties[k] = Backend.VALUE_REF + str(len(chunks))
+                    for i, chunk in enumerate(chunks):
+                        properties[self._chunk_property(k, i)] = chunk
+            else:
+                properties[k] = v
+        return properties
+
+    def _decode_metadata(self, f):
+        metadata = dict()
+        properties = f.get('properties', dict())
+        for k, v in properties.items():
+            if k.startswith('ref('):
+                continue
+            if v.startswith(Backend.VALUE_BASE64):
+                metadata[k] = base64.b64decode(v[len(Backend.VALUE_BASE64):])
+            elif v.startswith(Backend.VALUE_REF):
+                b64 = ''
+                for i in range(int(v[len(Backend.VALUE_REF):])):
+                    b64 += properties[self._chunk_property(k, i)]
+                metadata[k] = base64.b64decode(b64)
+            elif v.isdigit():
+                metadata[k] = int(v)
+            else:
+                metadata[k] = v
+        log.debug('{0}: metadata {1}'.format(f['name'], metadata))
+        return metadata
+
+    @property
+    @copy_ancestor_docstring
+    def has_native_rename(self):
+        return True
+
+    def __str__(self):
+        return 'gdrive folder %s@%s' % (self.login, self.prefix)
+
+    @retry
+    @copy_ancestor_docstring
+    def lookup(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)    
+        return self._decode_metadata(f)
+
+    @retry
+    @copy_ancestor_docstring
+    def get_size(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)
+        return int(f.get('size', u'0'))
+    @retry
+    @copy_ancestor_docstring
+    def open_read(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)  
+        return ObjectR(self.service, f, self._decode_metadata(f))
+
+    @prepend_ancestor_docstring
+    def open_write(self, key, metadata=None, is_compressed=False):
+        log.debug("key: {0}".format(key))
+        if metadata is None:
+            metadata = dict()
```

Comment:
If metadata is never `None`, why does _encode_metadata check for it?

Reply:
are you fully right never is None? if you are, could be removed

Reply:
No, what I'm saying is that here you ensure that it *metadata* will never be `None` from here on. But then you call another function (sorry, can't see the context at the moment, may be something like `_encode_metadata`) that checks if *metadata* is None and in that case returns `None` - defeating the check above.